File: src/main/java/simple/nodes/exps/BinExpNode.java
package simple.nodes.exps;

import simple.nodes.Node;

public abstract class BinExpNode implements Node {
    Node leftExp;
    Node rightExp;
    public  BinExpNode(Node leftNode, Node rightNode) {
        this.leftExp = leftNode;
        this.rightExp = rightNode;
    }
}

File: src/main/kotlin/interpreter/nodes/expressions/BinExpNode.kt
package interpreter.nodes.expressions

import exceptions.ParserException
import interpreter.Environment

class BinExpNode(val leftNode: Expression, val operation: String, val rightNode: Expression) : Expression {
    override fun execute(env: Environment): Any {
        val leftNodeResult = leftNode.execute(env) as? Int ?: TODO("cast to Int error")
        val rightNodeResult = rightNode.execute(env) as? Int ?: TODO("cast to Int error")

        return when(operation) {
            "+" -> leftNodeResult + rightNodeResult
            "-" -> leftNodeResult - rightNodeResult
            "*" -> leftNodeResult * rightNodeResult
            "==" -> leftNodeResult == rightNodeResult
            "!=" -> leftNodeResult != rightNodeResult
            ">" -> leftNodeResult > rightNodeResult
            "<" -> leftNodeResult < rightNodeResult
            ">=" -> leftNodeResult >= rightNodeResult
            "<=" -> leftNodeResult <= rightNodeResult

            else -> throw ParserException(operation)
        }
    }
}
